### ISSUE TITLE
Feature Request - A Way to Know if the State has Been Persisted Before

### ISSUE THREAD

**Issue by zeus82** [dotnet/orleans#6517]
I would be helpful to know if an `IPersistentState<T>` has been written before.  The scenario I'm thinking of is when your creating a new entity:
1. I want to know if that entity has been created before
2. There are actions that need to be performed when its being created for the first time
We could do something like this:
```
public interface IPersistentState<TState> where TState : new()
{
  TState State { get; set; }
  string Etag { get; }
  bool WasLoadedFromPersistence { get; }
  Task ClearStateAsync();
  Task WriteStateAsync();
  Task ReadStateAsync();
}
```

`WasLoadedFromPersistence` would be `false` the first time the state is loaded, and it would be forever `true` after you call `WriteStateAsync`.  Ideally, it would return to `false` when you call `ClearStateAsync`.

I understand this would be a breaking change because all the persistence providers would need to set it appropriately.

**Comment by ReubenBond** [comment]
Would checking that the ETag is not null/empty be sufficient, or perhaps adding a field (eg, CreationDate) to the TState?

**Comment by zeus82** [comment]
CreateDate could would.  The problem with using the ETag is that according to the [documentation](https://dotnet.github.io/orleans/Documentation/grains/grain_persistence/index.html#storage-provider-semantics), storage providers don't need to set it.

**Comment by zeus82** [comment]
I've implemented this feature locally, and I would like to put up a PR to have it reviewed.  What branch should the PR merge into?

This would be a breaking change for existing storage providers that aren't in this repo.

**Comment by EdeMeijer** [comment]
Just my 2 cents, I have a lot of 'entity' grains and I follow a pattern where each one has an `Exists` state field. Each entity needs to be created explicitly once with a `Create()` function, which does the initial setup and sets the flag to true. In every other function I assert that the `Exists` flag is true, and in the create function I assert that it is false. I have some base class for entity grains that handles most of this logic in a generic way.

**Comment by sergeybykov** [comment]
@zeus82 Does the approach with an explicit boolean flag would work for you? I'm personally hesitant about adding a flag like that as a standard feature because it would impose a tax of persisting that field on everyone.

**Comment by zeus82** [comment]
I was thinking that field could be computed by the provided when it goes to load the data from the backing store - all the providers in this repo are explicit on when they load the data from the store vs creating a new instance.  That was the only tax would really be the bool property added to the `IPersistentState` interface.

**Comment by sergeybykov** [comment]
Got it. Thanks. I don't see why this wouldn't work. 🙂

**Comment by zeus82** [comment]
Ok great... What branch should I commit these changes to?  Should I hold off until 4.0?

**Comment by sergeybykov** [comment]
Just create a branch in your fork, commit changes there, and submit a PR or post a link to your branch if not quite PR-ready. No need to wait for 4.0 for this I think.

**Comment by Galileon** [comment]
lol i am just dealing with the same problem :)

**Comment by zeus82** [comment]
This functionality has been merged and released in 3.3.0.  I think this issue can be closed.
